keps/sig-auth/20190730-oidc-discovery.md

```diff
@@ -195,10 +243,70 @@ configuration:
 ```
 
 The API server would treat these as `nonResourceURLs`, and restrict access
-appropriately. We will consider expanding `system:public-info-viewer` RBAC
-ClusterRole to grant access to the new paths; some other roles may have
-permission already via a pattern match on `nonResourceURLs` (e.g.
-`cluster-admin`).
+appropriately. We will provide a default RBAC `ClusterRole` called
+`service-account-issuer-discovery` that provides `GET` access to these
+`nonResourceURLs`, but will *not* provide a default `ClusterRoleBinding`. This
+leaves the decision of who is allowed to access these endpoints up to cluster
+admins. A default binding requires further discussion, including ongoing efforts
```

Comment:
While I am fine with this as a starting point, I am skeptical that this poses a security risk.  OIDC providers innately expose this type of configuration, and many do so on the public internet (Google, GitLab, Azure, etc).

To me, by not exposing this by default, we just make it hard for relying parties to consume this functionality agnostically of how the cluster was provisioned.  I think that misses the whole point of this KEP.

Reply:
While this endpoint appears to be fairly benign, we've thought the same thing about other permissions in public-info-viewer and been burned. Now benchmarks like CIS recommend blanket turning off anonymous auth which is an outcome but not a good one. In the future, we should maintain a very high bar when granting permissions to anonymous users.

Reply:
> While this endpoint appears to be fairly benign, we've thought the same thing about other permissions in public-info-viewer and been burned. Now benchmarks like CIS recommend blanket turning off anonymous auth which is an outcome but not a good one. In the future, we should maintain a very high bar when granting permissions to anonymous users.

I understand and respect the desire to limit what unauthenticated users can do.  But OIDC discovery endpoints are explicitly meant to be pre-authentication.  If we are careful in implementation (i.e. keep the endpoint "dumb" and have it just emit bytes with no user input), I think this is far safer than discovery endpoints that proxy for aggregated API servers.  Ideally we are talking about an endpoint only slightly more complex than `/healthz`.  Even if we include a default cluster role binding, the cluster admin can still reconcile protect it to tighten access.

If we want this functionality to be adopted by relying parties, it really needs to be present in all clusters by default.

Again, I am fine leaving this as-is and revisiting once the code matures.

Reply:
SG